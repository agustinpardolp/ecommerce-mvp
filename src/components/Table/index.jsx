import React from 'react'
import PropTypes from "prop-types";
import useWindowSize from '../../hooks/useWindowSize';
import TableDesktop from './components/TableDesktop';
import TableMobile from './components/TableMobile';

const Table = ({ columns, data, handleClick, btnLabel, handlePagination, paginationData, handleOrderColumns, loading }) => {
    const [isMobile] = useWindowSize();
    return (
        <>
            {isMobile ? (
                <TableMobile data={data} handleClick={handleClick} btnLabel={btnLabel} loading={loading} />
            ) : (
                <TableDesktop
                    loading={loading}
                    columns={columns} data={data}
                    handlePagination={handlePagination}
                    paginationData={paginationData}
                    handleOrderColumns={handleOrderColumns} />
            )}
        </>
    )
}
const Columns = PropTypes.shape({
    name: PropTypes.string,
    id: PropTypes.number,
    sort: PropTypes.string
});

const Data = PropTypes.shape({
    category: PropTypes.string,
    description: PropTypes.string,
    id: PropTypes.number,
    image: PropTypes.string,
    price: PropTypes.number,
    title: PropTypes.string,
    rating: PropTypes.shape({
        rate: PropTypes.number, count: PropTypes.number
    })
});

Table.propTypes = {
    columns: PropTypes.arrayOf(Columns).isRequired,
    data: PropTypes.arrayOf(Data),
    btnLabel: PropTypes.string,
    handleOrderColumns: PropTypes.func,
    handlePagination: PropTypes.func,
    loading: PropTypes.bool,
    paginationData: PropTypes.arrayOf(Data)
};

export default Table
