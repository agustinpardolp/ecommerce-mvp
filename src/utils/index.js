import { Provider } from 'react-redux';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import LanguageWrapper from '../components/LenguageWrapper';
import { store } from "../store";

const DEBOUNCE_DELAY_TIME = 200;
export const debounce = (func, timeout = DEBOUNCE_DELAY_TIME) => {
  let timer;
  return (...args) => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      func(...args);
    }, timeout);
  };
};

export const getFullWrappedComponent = (
  Component,
  props = {},
  reduxStore = store,
) => (
  <BrowserRouter>
    <MemoryRouter>
      <LanguageWrapper>
        <Provider store={reduxStore}>
          <Component {...props} />
        </Provider>
      </LanguageWrapper>
    </MemoryRouter>
  </BrowserRouter>
);


export const handleSizeChange = () => {
  global.innerWidth = 400;
  global.dispatchEvent(new Event('resize'));
};
