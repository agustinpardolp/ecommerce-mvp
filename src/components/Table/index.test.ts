import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "@testing-library/react";
import Table from "./index";
import { getFullWrappedComponent, handleSizeChange } from "../../utils";

describe("Render Table with no errors", () => {
  const mockProps = {
    columns: [
      { name: "Test.Title Column", id: 1 },
      { name: "Test.Category Column", id: 2 },
      { name: "Test.Price Column", id: 3 },
    ],
    data: [
      {
        category: "men's clothing",
        description: "",
        id: 1,
        image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
        price: 109.95,
        rating: { rate: 3.9, count: 120 },
        title: "Product test 1",
      },
    ],
    handleClick: jest.fn(),
    btnLabel: "",
    handlePagination: jest.fn(),
    paginationData: [],
    handleOrderColumns: jest.fn(),
  };

  test("Render Table items correctly", () => {
    const { container } = render(getFullWrappedComponent(Table, mockProps));
    expect(screen.getAllByText("men's clothing")[0]).toBeDefined();
    expect(screen.getByText("Product test 1")).toBeDefined();
    expect(screen.getByText(109.95)).toBeDefined();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("Correctly displays empty list message", () => {
    const noDataProps = {
      columns: [
        { name: "Test.Title Column", id: 1 },
        { name: "Test.Category Column", id: 2 },
        { name: "Test.Price Column", id: 3 },
      ],
      data: [],
      handleClick: jest.fn(),
      btnLabel: "",
      handlePagination: jest.fn(),
      paginationData: [],
      handleOrderColumns: jest.fn(),
    };
    render(getFullWrappedComponent(Table, noDataProps));
    expect(screen.getByText("No se encontraron datos")).toBeInTheDocument();
  });

  test("Render correctly mobile version", async () => {
    handleSizeChange();
    render(getFullWrappedComponent(Table, mockProps));
    const mobileCard = await screen.findAllByTestId("mobile-card");
    expect(mobileCard[0]).toBeInTheDocument();
  });
});
