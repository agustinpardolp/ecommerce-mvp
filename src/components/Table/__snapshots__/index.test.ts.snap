// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Table with no errors Render Table items correctly 1`] = `
<div
  class="sc-gsDKAQ YHamg"
>
  <table
    class="ui celled fixed sortable table"
  >
    <thead
      class=""
    >
      <tr
        class=""
      >
        <th
          class=""
        >
          Test.Title Column
        </th>
        <th
          class=""
        >
          Test.Category Column
        </th>
        <th
          class=""
        >
          Test.Price Column
        </th>
      </tr>
    </thead>
    <tbody
      class=""
    >
      <tr
        class=""
      >
        <td
          class=""
        >
          <h4
            class="ui image header"
          >
            <img
              class="ui medium rounded image"
              src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
            />
            <div
              class="content"
            >
              Product test 1
              <div
                class="sub header"
              >
                men's clothing
              </div>
            </div>
          </h4>
        </td>
        <td
          class=""
        >
          men's clothing
        </td>
        <td
          class=""
        >
          109.95
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="sc-bdvvtL dGibiM"
  />
</div>
`;

exports[`Render Table with no errors Render Tableitems correctly 1`] = `
<div
  class="sc-gsDKAQ YHamg"
>
  <table
    class="ui celled fixed sortable table"
  >
    <thead
      class=""
    >
      <tr
        class=""
      >
        <th
          class=""
        >
          Test.Title Column
        </th>
        <th
          class=""
        >
          Test.Category Column
        </th>
        <th
          class=""
        >
          Test.Price Column
        </th>
      </tr>
    </thead>
    <tbody
      class=""
    >
      <tr
        class=""
      >
        <td
          class=""
        >
          <h4
            class="ui image header"
          >
            <img
              class="ui medium rounded image"
              src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
            />
            <div
              class="content"
            >
              Product test 1
              <div
                class="sub header"
              >
                men's clothing
              </div>
            </div>
          </h4>
        </td>
        <td
          class=""
        >
          men's clothing
        </td>
        <td
          class=""
        >
          109.95
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="sc-bdvvtL dGibiM"
  />
</div>
`;
